maintainers/scripts/haskell/hydra-report.hs

```diff
@@ -0,0 +1,233 @@
+         , unfinishedBuilds = unfinishedBuilds lhs + unfinishedBuilds rhs
+         , canceledBuilds = canceledBuilds lhs + canceledBuilds rhs
+         }
+
+instance Monoid StatusSummary where
+   mempty = StatusSummary 0 0 0 mempty mempty mempty mempty mempty mempty
+
+main :: IO ()
+main = do
+   args <- getArgs
+   case args of
+      ["get-report"] -> getBuildReports
+      ["ping-maintainers"] -> printMaintainerPing
+      ["mark-broken-list"] -> printMarkBrokenList
+      _ -> putStrLn "Usage: get-report | ping-maintainers | mark-broken-list"
+
+reportFileName = "haskell-updates-build-report.json"
+
+showT :: Show a => a -> Text
+showT = Text.pack . show
+
+getBuildReports :: IO ()
+getBuildReports = runReq defaultHttpConfig $ do
+   -- GET request http response
+   now <- liftIO getCurrentTime
+   r <- req GET (https "hydra.nixos.org" /: "jobset" /: "nixpkgs" /: "haskell-updates" /: "evals") NoReqBody jsonResponse mempty
+   let eval = Seq.lookup 0 . evals $ (responseBody r :: JobsetEvals)
+   eval & maybe (liftIO $ putStrLn "No Evalution found") \eval -> do
+      (buildReports :: Seq Build) <- forM (builds eval) \buildId ->
+         responseBody <$> req GET (https "hydra.nixos.org" /: "build" /: showT buildId) NoReqBody jsonResponse mempty
+      liftIO $ encodeFile reportFileName (eval, now, buildReports)
+
+maintainersQuery = "with import ./. {}; lib.mapAttrs (_: v: map (m: m.github) v.meta.maintainers) (lib.filterAttrs (_: v: builtins.length (v.meta.maintainers or []) > 0) haskellPackages)"
+command = "nix-instantiate"
+params = ["--eval", "--strict", "--json", "-"]
+
+type MaintainerMap = Map Text (Seq Text)
+
+getMaintainerMap :: IO MaintainerMap
+getMaintainerMap = do
+   nixOutput <- readProcess command params maintainersQuery
+   pure $ fromMaybe (error "Failed to decode MaintainerMap") (decodeStrict' (ByteString.pack nixOutput))
+
+buildSummary :: MaintainerMap -> Seq Build -> StatusSummary
+buildSummary maintainerMap = foldMap toSummary
+  where
+   toSummary Build{finished, buildstatus, job, id} =
+      case (finished, buildstatus, packageName, maintainers) of
+         (0, _, _, _) -> mempty{unfinishedBuilds = 1}
+         (_, 0, _, _) -> mempty{successfulBuilds = 1}
+         (_, 7, _, _) -> mempty{canceledBuilds = 1}
+         (_, 2, Just n, Just m) -> mempty{failedMaintainedDependencies = Seq.singleton (BuildName n id, m)}
+         (_, 2, Just n, _) -> mempty{failedPackageDependencies = Seq.singleton (BuildName n id)}
+         (_, 2, _, _) -> mempty{failedJobDependencies = Seq.singleton (BuildName jobName id)}
+         (_, _, Just n, Just m) -> mempty{failedMaintainedBuilds = Seq.singleton (BuildName n id, m)}
+         (_, _, Just n, _) -> mempty{failedPackageBuilds = Seq.singleton (BuildName n id)}
+         _ -> mempty{failedJobBuilds = Seq.singleton (BuildName jobName id)}
+     where
+      jobName = fromMaybe job (Text.stripSuffix ".x86_64-linux" job)
```

Comment:
Note that I'm still planning to enable `aarch64` and `x86_64-darwin`.

Reply:
Yeah, I know. This script will at the moment just throw all the jobs from other platforms into "other jobs". Which is okay for now, maybe? Anyways I am not sure what the best desired behavior in the presence of multiple platforms is. Probably we want to somehow group builds for one package together? But the usability will depend a lot on how many failures we will experience.

Reply:
The updated version of the script takes this into account.